garrysmod/lua/vgui/dcombobox.lua

```diff
@@ -1,289 +1,355 @@
+
+Derma_Hook( PANEL, "Paint", "Paint", "ComboBox" )
+
+Derma_Install_Convar_Functions( PANEL )
+
+AccessorFunc( PANEL, "m_bDoSort", "SortItems", FORCE_BOOL )
+
+function PANEL:Init()
+
+	-- Create data tables
+	self.Choices = {}
+	self.Data = {}
+	self.ChoiceIcons = {}
+	self.Spacers = {}
+
+	-- Create button
+	self.DropButton = vgui.Create( "DPanel", self )
+	self.DropButton.Paint = function( panel, w, h ) derma.SkinHook( "Paint", "ComboDownArrow", panel, w, h ) end
+	self.DropButton:SetMouseInputEnabled( false )
+	self.DropButton.ComboBox = self
+
+	-- Apply default panel hight
+	self:SetTall( 22 )
+
+	-- Setup internals
+	self:SetContentAlignment( 4 )
+	self:SetTextInset( 8, 0 )
+	self:SetIsMenu( true )
+	self:SetSortItems( true )
+
+end
+
+function PANEL:CloseMenu()
+
+	if ( IsValid( self.Menu ) ) then
+		self.Menu:Remove()
+	end
+
+	self.Menu = nil
+
+end
+
+function PANEL:Clear()
+
+	-- Does not allocate another table
+	self:SetText( "" )
+	table.Empty( self.Choices )
+	table.Empty( self.Data )
+	table.Empty( self.ChoiceIcons )
+	table.Empty( self.Spacers )
+	self.selected = nil
+	self:CloseMenu()
+
+end
+
+function PANEL:GetSelectedID()
+
+	return self.selected
+
+end
+
+function PANEL:GetChoiceCount()
+
+	return #self.Choices
+
+end
+
+function PANEL:GetOptionText( id )
+
+	return self.Choices[ id ]
+
+end
+
+function PANEL:GetOptionData( id )
+
+	return self.Data[ id ]
+
+end
+
+function PANEL:GetOptionIcon( id )
+
+	return self.ChoiceIcons[ id ]
+
+end
+
+function PANEL:GetOptionSpacer( id )
+
+	return self.Spacers[ id ]
+
+end
+
+function PANEL:AddSpacer()
+
+	local id = self:GetChoiceCount()
+
+	self.Spacers[ id ] = true
+
+end
+
+function PANEL:GetOptionTextByData( data )
+
+	for id, dat in pairs( self.Data ) do
+		if ( dat == data ) then
+			return self:GetOptionText( id )
+		end
+	end
+
+	-- Try interpreting it as a number
+	for id, dat in pairs( self.Data ) do
+		if ( dat == tonumber( data ) ) then
+			return self:GetOptionText( id )
+		end
+	end
+
+	-- In case we fail
+	return data
+
+end
+
+function PANEL:PerformLayout()
+
+	self.DropButton:SetSize( 15, 15 )
+	self.DropButton:AlignRight( 4 )
+	self.DropButton:CenterVertical()
+
+	-- Make sure the text color is updated
+	DButton.PerformLayout( self, w, h )
+
+end
+
+function PANEL:ChooseOption( value, id )
+
+	self:CloseMenu()
+	self:SetText( value )
+
+	-- This should really be the here, but it is too late now and convar
+	-- changes are handled differently by different child elements
+	-- self:ConVarChanged( self.Data[ id ] )
+
+	self.selected = id
+	self:OnSelect( id, value, self.Data[ id ] )
+
+end
+
+function PANEL:ChooseOptionID( id )
+
+	local value = self:GetOptionText( id )
+	self:ChooseOption( value, id )
+
+end
+
+function PANEL:RemoveChoiceID( id )
+
+	-- The second argument must be convertable to number
+	-- Removing non-positive or fractional does nothing
+	-- Entry will be removed only on positive integers
+	local index = tonumber(id) or self:GetChoiceCount()

```

Comment:
The function should not be falling back to any ID, it should just exit with invalid input.

I am also unsure whether these remove functions should even exist.

Reply:
Fixed in f86f760e49b9d9d75302988738f5852bffac5c95